File: app/src/main/java/com/example/myapplication/CoroutineUtils.java
package com.example.myapplication;

import android.util.Log;

import org.jetbrains.annotations.NotNull;

import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.FutureTask;

public class CoroutineUtils {
    public static <T> CoroutinePro submit(CorouRunnable<T> runnable, Callback<T> callback) {
        CoroutinePro innerCorou = new CoroutinePro<String, String, T>() {

            @Override
            protected T doInBackground(String... args) {
                Log.e("xxx", String.format("start doInBackground: on Thread [%s]", Thread.currentThread().getName()));
//                runnable.run();
                FutureTask<T> futureTask = new FutureTask<>(runnable);
                futureTask.run();
                try {
                    return futureTask.get();
                } catch (ExecutionException e) {
                    e.printStackTrace();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                return null;
            }

            @Override
            protected void onProgressUpdate(@NotNull String[] data) {
                super.onProgressUpdate(data);
            }

            @Override
            protected void onPostExecute(T res) {
                super.onPostExecute(res);
                Log.e("xxx", String.format("receive res on thread [%s],res = %s", Thread.currentThread().getName(), res));
                callback.onFinish(res);
            }

            @Override
            protected void onCancelled() {
                super.onCancelled();
                callback.onCancel();
            }

            @Override
            protected void onError(@NotNull Exception e) {
                super.onError(e);
                callback.onError();
            }
        }.executeOnIO(runnable.getTag());
        return innerCorou;
    }

    public interface Callback<T> {
        void onFinish(T res);

        void onError();

        void onCancel();
    }

}

File: app/src/main/java/com/example/myapplication/CoroutineUtils.kt
package com.example.myapplication

import android.util.Log
import kotlinx.coroutines.*
import java.util.concurrent.ExecutionException
import java.util.concurrent.FutureTask

object CoroutineUtils {

    val tag = "CoroutineUtils"
    /**
     *  替换回调的写法,异步执行，回调处理
     */
    @JvmStatic
    fun <R> waitAndExcuAsyncDefault(backgroundFunc: () -> R, mainFunc: (R?) -> Unit) {
        GlobalScope.launch(Dispatchers.Main) {
            var result: R? = null
            withContext(Dispatchers.IO) {
                try {
                    log(backgroundFunc.toString())
                    result = backgroundFunc()
                } catch (e: Exception) {

                }
                result
            }
            log(mainFunc.toString())
            mainFunc(result)
        }
    }

    @JvmStatic
    fun <T> excuOnIO(callable: CorouRunnable<T>) {
        val scope = CoroutineScope(Dispatchers.IO)
        scope.launch {
            log("excuOnIO: task = " + callable.mTag)
            callable.call()
        }
    }

    @JvmStatic
    fun <T> excuOnIOWithCallback(callable: CorouRunnable<T>,callback : Callback<T>) {
        val scope = CoroutineScope(Dispatchers.Main)
        scope.launch {
            var result : T? = null
            withContext(Dispatchers.IO){
                try {
                    log("excuOnIOWithCallback: task = " + callable.mTag)
                    result = callable.call()
                }catch (e : Exception){
                    callback.onError(e.message)
                }
                result
            }
            callback.onFinish(result)
        }
    }

    @InternalCoroutinesApi
    @JvmStatic
    fun <T> submitWithCallback(runnable: CorouRunnable<T>, callback: Callback<T?>): CoroutinePro<*, *, *> {
        return object : CoroutinePro<String?, String?, T>() {
            override fun doInBackground(vararg args: String?): T? {
                val futureTask = FutureTask(runnable)
                futureTask.run()
                try {
                    return futureTask.get()
                } catch (e: ExecutionException) {
                    e.printStackTrace()
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
                return null
            }

            override fun onPostExecute(res: T?) {
                super.onPostExecute(res)
                callback.onFinish(res)
            }

            override fun onCancelled() {
                super.onCancelled()
                callback.onCancel()
            }

            override fun onError(e: Exception) {
                super.onError(e)
                callback.onError(e.message)
            }
        }.executeOnIO(runnable.tag)
    }

    @InternalCoroutinesApi
    @JvmStatic
    fun <T> submit(runnable: CorouRunnable<T>): CoroutinePro<*, *, *> {
        return object : CoroutinePro<String?, String?, T>() {
            override fun doInBackground(vararg args: String?): T? {
                val futureTask = FutureTask(runnable)
                futureTask.run()
                try {
                    return futureTask.get()
                } catch (e: ExecutionException) {
                    e.printStackTrace()
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
                return null
            }

            override fun onPostExecute(res: T?) {
                super.onPostExecute(res)
            }

        }.executeOnIO(runnable.tag)
    }

    interface Callback<T> {
        fun onFinish(res: T?)
        fun onError(mes : String?)
        fun onCancel()
    }

    fun log(mes : String?){
        val shouldLog= true
        if (shouldLog){
            Log.e(tag, String.format("message = %s ,t = %s", mes,Thread.currentThread().name))
        }
    }
}